src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -18,38 +18,100 @@
 import io.supertokens.pluginInterface.oauth.OAuthStorage;
 
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+
 public class OAuth {
 
-    public static void getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
-                                           String redirectURI, String responseType, String scope, String state)
-            throws InvalidConfigException {
-        // TODO:
-        // - validate that client_id is present for this tenant
-        // - call hydra
-        // - if location header is:
-        //     - localhost:3000, then we redirect to apiDomain
-        //     - public url for hydra, then we throw a 400 error with the right json
-        //     - else we redirect back to the client
+    private static final String LOCATION_HEADER_NAME = "Location";
+    private static final String COOKIES_HEADER_NAME = "Set-Cookie";
+    private static final String ERROR_LITERAL = "error=";
+    private static final String ERROR_DESCRIPTION_LITERAL = "error_description=";
+
+
+    public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
+                                                        String redirectURI, String responseType, String scope, String state)
+            throws InvalidConfigException, HttpResponseException, IOException, OAuthAuthException, StorageQueryException,
+            TenantOrAppNotFoundException {
 
         OAuthStorage oauthStorage = StorageUtils.getOAuthStorage(storage);
 
         String redirectTo = null;
+        List<String> cookies = null;
+
+        String publicOAuthProviderServiceUrl = Config.getConfig(appIdentifier.getAsPublicTenantIdentifier(), main).getOAuthProviderPublicServiceUrl();
 
         if (!oauthStorage.doesClientIdExistForThisApp(appIdentifier, clientId)) {
-            redirectTo = Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl() +
-                    "/oauth2/fallbacks/error?error=invalid_client&error_description=Client+authentication+failed+%28e" +
-                    ".g.%2C+unknown+client%2C+no+client+authentication+included%2C+or+unsupported+authentication" +
-                    "+method%29.+The+requested+OAuth+2.0+Client+does+not+exist.";
+            throw new OAuthAuthException("invalid_client", "Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). The requested OAuth 2.0 Client does not exist.");
         } else {
             // we query hydra
+            Map<String, String> queryParamsForHydra = constructHydraRequestParamsForAuthorizationGETAPICall(clientId, redirectURI, responseType, scope, state);
+            Map<String, String> responseHeaders = new HashMap<>();
+
+            //TODO maybe check response status code? Have to modify sendGetRequest.. for that
```

Comment:
In general, if the response code is < 400, the function doesn't throw, else it throws. I think based on the API spec, it would return a 302, or then >= 400 (but this needs to be tested properly). If it's true, we dont need to change the function.

Reply:
I agree that we have to first test it.

Reply:
Keeping this comment unresolved cause we have to test it first. If you have already done so, please resolve this comment.

Reply:
I couldn't find any cases where it went above 400. (Generally, 302 or 303). Of course, this doesn't prove that there are _no cases_ where it would return a 400 or above, but I couldn't find any.

Reply:
In what case does it return a 400?

Reply:
I couldn't find any. I was referencing 400 because the HttpRequest throws exception above 400.